### PR TITLE
Contact and Account duplicate checks

### PR DESCRIPTION
**Pull request by ptrelease**
Introduced checks to avoid upserting a Contact record if one already exists in Salesforce.

Introduced checks to avoid upserting an Account record if one already exists in Salesforce.

### CODE REVIEW

File: lib/apis/salesforce/salesforce_api.rb [hunk 1]
```diff
@@ -736,6 +796,49 @@ def get_salesforce_expression_of_interest_reference(expression_of_interest)
...
+
+      rescue Restforce::NotFoundError
+        Rails.logger.info("Unable to find contact with external id #{user.id} " \
+          "will attempt to find contact using a name and Email match")  
+      end
+      
+      unless contact_salesforce_id 
+
+        contact_collection_from_salesforce = 
+          @client.query_all("select Id from Contact where name = '#{user.name}' and Email = '#{user.email}' ")
```
Comment (stuartmccoll) [inline lib/apis/salesforce/salesforce_api.rb:829]:
```suggestion
          @client.query_all("select Id from Contact where name = '#{user.name}' and Email = '#{user.email}'")
```

File: lib/apis/salesforce/salesforce_api.rb [hunk 2]
```diff
@@ -550,15 +584,41 @@ def upsert_contact_in_salesforce(user, salesforce_organisation_reference)
...
-         "Created a Contact record in Salesforce with reference: #{salesforce_contact_reference}"
-        )
-
-       salesforce_contact_reference
-
-    end
+    # Upserts to a Contact record in Salesforce using the User instance's id
+    # Removes the FirstName, MiddleName, Suffix attributes, and 
+    # populates LastName with the User.name value from Funding Frontend
+    # Retries accounted for in outer scope call.
```
Comment (stuartmccoll) [inline lib/apis/salesforce/salesforce_api.rb:593]: Should this read as `Calling function should handle exceptions/retries`? That way it's clear to anybody attempting to re-use this method elsewhere that they need to account for this.

File: lib/apis/salesforce/salesforce_api.rb [hunk 3]
```diff
@@ -523,19 +523,53 @@ def create_organisation_in_salesforce(organisation)
 
     end
 
-    # Method to upsert a Salesforce Contact record. Retries accounted for in outer scope call.
+    # Method to orchestrate upserting a Salesforce Contact record. 
+    # Tries to find an existing Contact record in Salesforce first.
+    # Then calls an appropriate upsert.  Upserts by Salesforce's
+    # Contact Id if known.  Otherwise upserts by the User objects id.
+    # Retries accounted for in outer scope call.
```
Comment (stuartmccoll) [inline lib/apis/salesforce/salesforce_api.rb:530]: Should this read as `Calling function should handle exceptions/retries`? That way it's clear to anybody attempting to re-use this method elsewhere that they need to account for this.

**Review by stuartmccoll — CHANGES_REQUESTED** [review]

File: lib/apis/salesforce/salesforce_api.rb [hunk 4]
```diff
@@ -736,6 +796,49 @@ def get_salesforce_expression_of_interest_reference(expression_of_interest)
...
+    # If no match found, then tries to find a Contact record with a matching name/email
+    # Combination.  
+    # A Salesforce Id for the Contact record is returned if a match is made.  Otherwise nil.
+    # Exception handling and retries accounted for in outer scope
+    #
+    # @param [User] user An instance of User which is the current user
+    #
+    # @return [String] contact_salesforce_id A string representing Salesforce Id 
+    #                                        for the Contact record, or nil
+    def find_an_existing_contact_matching_this_user(user)
```
[the remark this answers is not in the document]
Comment (ptrelease) [reply lib/apis/salesforce/salesforce_api.rb:811]: find_matching_contact_for_user is fine.  Will change.

File: lib/apis/salesforce/salesforce_api.rb [hunk 5]
```diff
@@ -736,6 +796,49 @@ def get_salesforce_expression_of_interest_reference(expression_of_interest)
 
     end
 
+
+    # Method check Salesforce for existing Contact records for the passed User instance
+    # Firstly checks if a Contact record exists with an external ID matching the user.id
+    # If no match found, then tries to find a Contact record with a matching name/email
+    # Combination.  
+    # A Salesforce Id for the Contact record is returned if a match is made.  Otherwise nil.
+    # Exception handling and retries accounted for in outer scope
```
[the remark this answers is not in the document]
Comment (ptrelease) [reply lib/apis/salesforce/salesforce_api.rb:805]: like it

File: lib/apis/salesforce/salesforce_api.rb (hunk 1 above)
Comment (ptrelease) [reply lib/apis/salesforce/salesforce_api.rb:829]: I thought you'd prefer the space here ;-)

File: lib/apis/salesforce/salesforce_api.rb (hunk 3 above)
Comment (ptrelease) [reply lib/apis/salesforce/salesforce_api.rb:530]: Changed

**Comment by paulmsmith** [comment]
@ptrelease Github tip: Did you know you can create WIP pull requests now? I think github calls them a 'draft' and they can't be merged until you say so.

**Comment by ptrelease** [comment]
> @ptrelease Github tip: Did you know you can create WIP pull requests now? I think github calls them a 'draft' and they can't be merged until you say so.

File: lib/apis/salesforce/salesforce_api.rb [hunk 6]
```diff
@@ -492,14 +492,37 @@ def update_agrees_to_user_research(user)
...
+    end
+
+    # Upserts to an Account record in Salesforce using the organisation.id
+    # Calling function should handle exceptions/retries
+    #
+    # @param [Organisation] organisation An instance of a Organisation object
+    #
+    # @return [String] salesforce_account_id A Salesforce Account Id for the Organisation
+    def upsert_account_by_organisation_id(organisation)
+      salesforce_account_id = @client.upsert!(
```
Comment (stuartmccoll) [inline lib/apis/salesforce/salesforce_api.rb:525]: We don't need to assign to a variable here, as we're just returning what the `upsert` call returns.

File: lib/apis/salesforce/salesforce_api.rb [hunk 7]
```diff
@@ -514,28 +537,118 @@ def create_organisation_in_salesforce(organisation)
...
+    # Upserts to an Account record in Salesforce using the salesforce Account Id
+    # Calling function should handle exceptions/retries
+    #
+    # @param [Organisation] organisation An instance of a Organisation object
+    # @param [String] salesforce_account_id A salesforce Account Id 
+    #                                       for the User's organisation
+    #
+    # @return [String] salesforce_account_id A Salesforce Account Id for the Organisation
+    def upsert_account_by_salesforce_id(organisation, salesforce_account_id)
+      salesforce_account_id = @client.upsert!(
```
Comment (stuartmccoll) [inline lib/apis/salesforce/salesforce_api.rb:551]: We don't need to assign to a variable here, as we're just returning what the `upsert` call returns.

**Review by stuartmccoll — APPROVED** [review]

**Pull request merged**